Added Trick manager

Sp4tz7 opened the pull request:
Close #54

Sp4tz7 committed:
Added Trick manager

Jduret commented on src/Service/Trick/Manager/TrickManager.php:
```diff
@@ -0,0 +1,98 @@
+    {
+        $trick->setSlug($this->slugManager->slugThis($trick->getName()));
+
+        $imageFile = $form['image']->getData();
```
Je pense qu'il serait plus judicieux de faire passer les "data" en paramètre plutot que de lier ton persist à un objet formulaire

Sp4tz7 replied:
Ok, bonne idée. Merci je corrige.

Jduret commented on src/Service/Trick/Manager/TrickManager.php:
```diff
@@ -0,0 +1,98 @@
+
+        if ($imageFile) {
+            $filename = $this->fileUploader->upload($imageFile);
+            $trick->setImage($filename);
```
Quel différence entre `setImage` et `addImage` ? Les 2 manipulent les mêmes images ou `setImage` manipule l'image principal du trick et `addImage` gère les autres ?

Sp4tz7 replied:
SetImage() définit l'image principale
SetImages() définit une collection d'images à l'objet Trick
addImage() ajoute un objet image à la collection images de l'objet Trick

Jduret replied:
Alors dans ce cas, le nommage n'est pas très clair. Je te mets une idée mais tu peux trouver mieux surement :
- setMainImage
- setImages (si la main image correspond à l'image 0 de ta collection)
- setRemainingImages (ou quelque chose du genre si la main image n'est pas dedans)
- addImage ou addRemainingImage

En espérant que tu auras compris l'idée

Sp4tz7 replied:
Oui, effectivement. En fait c'est le make:entity de symfony qui a créé automatiquement ces fonctions et je n'ai pas pensé à les changer.

Jduret commented on src/Service/Trick/Manager/TrickManager.php:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        // remove image if deleted
+        foreach ($this->originalImages as $key => $originalImage) {
```
FYI : pas besoin de faire `$key=>$originalImage` si `$key` ne te sers à rien

Sp4tz7 replied:
C'est juste... je l'avais utilisé avant puis supprimé. J'ai juste oublié le paramètre.

je corrige...

Jduret commented on src/Service/Trick/Manager/TrickManager.php:
```diff
@@ -0,0 +1,98 @@
+            if (false === $trick->getImages()->contains($image)) {
+                $this->fileUploader->remove($image);
+            }
+        }
```
Faire ça dans une méthode dédiée
```php
/** 
 * @description Permet de filtrer les images supprimées puis de la supprimer
 * @return array La liste des images supprimées
 */
protected function removeDeletedImages($originalImages, $trickImages)
{
    return array_map([$this, "removeImage"], array_filter(
        $originalImages,
        //array_filter ne garder que les élements où la fonction retourne VRAI
        function($imageData) use($trickImages) {
            return false === $trickImages->contains($imageData['image_object']);
        }
    ));
}

protected function removeImage($imageData)
{
    $this->fileUploader->remove($imageData['image_object']);
    return $imageData;
}
```

Attention, j'ai écrit ça directement dans github pour montrer une idée basé sur le "fonctional programming". Il peut contenir des erreurs

Sp4tz7 replied:
Ok. Je verrai pour implémenter cette idée dans une prochaine release.

Jduret commented on src/Form/ImageType.php:
```diff
@@ -18,10 +18,22 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
         ;
+        $builder->get('file_name')->addModelTransformer(new CallbackTransformer(
+            function ($TEST) {
```
Je suis preneur d'une explication de cette variable `$TEST`

Sp4tz7 replied:
C'est une coquille 🤣

Jduret reviewed (changes requested):
Beau travail de refactorisation du controller. 
Par contre je ne vois aucun test unitaire de ton TrickManager... 🤔

Commit:
Fix PR 55 reviews:

- Passed the data as function parameters instead of calling them in the
  object
- Removec test variable
- Removed unused $key variable
